lib/command.ts

```diff
@@ -74,18 +83,29 @@ export default abstract class BalenaCommand extends Command {
 		await (await import('./utils/patterns')).checkLoggedIn();
 	}
 
+	/**
+	 * Read stdin contents and make available to command.
+	 *
+	 * This approach could be improved in the future to automatically set argument
+	 * values from stdin based in configuration, minimising command implementation.
+	 */
+	protected async getStdin() {
+		this.stdin = await (await import('get-stdin'))();
+	}
+
 	protected async init() {
-		const requireElevatedPrivileges = (this.constructor as typeof BalenaCommand)
-			.root;
-		const requireAuthentication = (this.constructor as typeof BalenaCommand)
-			.authenticated;
+		const ctr = this.constructor as typeof BalenaCommand;
 
-		if (requireElevatedPrivileges) {
+		if (ctr.root) {
 			await BalenaCommand.checkElevatedPrivileges();
 		}
 
-		if (requireAuthentication) {
+		if (ctr.authenticated) {
 			await BalenaCommand.checkLoggedIn();
 		}
+
+		if (ctr.readStdin) {
+			await this.getStdin();
```

Comment:
hmm but stdin should only be read if the `notes` argument is not provided, no? Apparently the `key add` command conditionally reads from `stdin`:
https://github.com/balena-io/balena-cli/blob/v11.30.6/lib/actions/keys.ts#L99-L116

> I've added piped input support, using the same library that capitano uses

Does Capitano read from stdin regardless or whether the argument was provided? If there are no side effects then I suppose it is OK. Things that come mind are interactive menus, or a password prompt.

> Now tested working on Windows cmd, powershell, msys

And Linux? :-)  (msys uses a CLI release for Windows...)

Reply:
> hmm but stdin should only be read if the notes argument is not provided, no?

> Does Capitano read from stdin regardless or whether the argument was provided?

I'm not sure about capitano.  Can you elaborate on a situtation you are worried about?  I don't see reading the contents as harmful, whether or not we use them later.  If we wanted to do it conditionally we would need to have the base class responsible for triggering the parsing of commands (I plan to try to do this, but I wanted to keep things simple first, so that we can make some progress without getting bogged down).  (The fact that oclif relies on the concrete implemetation to call the parse method seems like a strange design choice and really limiting to me).

> And Linux? :-) (msys uses a CLI release for Windows...)

Yep, I always test on linux (but should have been clearer).

Reply:
> Things that come mind are interactive menus, or a password prompt.
>> Can you elaborate on a situtation you are worried about?

Basically whether there'd be side effects that prevented something else from working. In #1556 we added `process.stdin.pause()` before the CLI exits, as otherwise the CLI was hanging trying to read from stding -- so that kind of situation (maybe it was Capitano's behaviour that required that line to be added, and I didn't even know about it...) But as you tested it and it works, it should be fine. Approving the PR. 👍

Reply:
Ok thanks, I'll keep an eye open for that! (I haven't managed to provoke anything strange in tests yet)